gms/gossiper.cc

```diff
@@ -757,11 +757,12 @@ future<> gossiper::do_status_check() {
 
         // check for dead state removal
         auto expire_time = get_expire_time_for_endpoint(endpoint);
-        const auto host_id = get_host_id(endpoint);
-        if (!is_alive && (now > expire_time)
-             && (!get_token_metadata_ptr()->is_normal_token_owner(host_id))) {
+        if (!is_alive && (now > expire_time)) {
+          const auto host_id = get_host_id(*eps);
+          if (!host_id || !get_token_metadata_ptr()->is_normal_token_owner(host_id)) {
```

Comment:
I'm worried that https://github.com/scylladb/scylladb/issues/16936 indicates a real bug that we have, and this change would hide it.

The bug would be that an endpoint appears in gossiper state which does not have a Host ID.
My question is how can this happen?

I thought it's an invariant we preserve that every endpoint comes with a Host ID.
Whenever node starts gossiping, it always inserts its host ID.
And also after a full cluster restart, when nodes are gossiping their peers' state restored from disk, they also include their host IDs.

So isn't there some larger problem we're dealing with? That host IDs are disappearing?

Reply:
We can add an internal error in non-release modes to help diagnose the problem in testing and follow-through, as suggested here, in release mode so not to interfere with production.